status/api/wallet.nim

```diff
@@ -25,82 +27,106 @@ type
+    address = try: acct.address.parseAddress
+              except ValueError as e: return WalletAccountResult.err "Error " &
+                "parsing address: " & e.msg
     publicKey = walletPubKeyResult.get.some
 
   return self.storeWalletAccount(name, address, publicKey, accountType, path)
 
 proc addWalletAccount*(self: StatusObject, name, password,
-  dir: string): WalletAccountResult {.raises: [Defect, Exception, ref IOError,
-  ref OSError, SqliteError, UserDbError, ref ValueError].} =
+  dir: string): WalletAccountResult =
 
   if not self.isLoggedIn:
     return WalletAccountResult.err "Not logged in. You must be logged in to " &
       "create a new wallet account."
 
-  let address = self.userDb.getSetting(string, SettingsCol.WalletRootAddress)
-  if address.isNone:
-    return WalletAccountResult.err "Unable to get wallet root address from " &
-      "settings. Cannot create a derived wallet."
-
-  let
-    lastDerivedPathIdx =
-      self.userDb.getSetting(int, SettingsCol.LatestDerivedPath, 0)
-    loadedAccount = ?self.accountsGenerator.loadAccount(
-      address.get.parseAddress, password, dir)
-    newIdx = lastDerivedPathIdx + 1
-    path = fmt"{PATH_WALLET_ROOT}/{newIdx}"
-    walletAccount = ?self.storeDerivedAccount(loadedAccount.id, KeyPath path,
-      name, password, dir, AccountType.Generated)
-
-  self.userDb.saveSetting(SettingsCol.LatestDerivedPath, newIdx)
+  const
+    errorMsgSetting = "Error getting setting: "
+    errorMsgParse = "Error parsing root wallet address: "
+  try:
+    let address = self.userDb.getSetting(string, SettingsCol.WalletRootAddress)
+    if address.isNone:
+      return WalletAccountResult.err "Unable to get wallet root address from " &
+        "settings. Cannot create a derived wallet."
 
-  WalletAccountResult.ok(walletAccount)
+    let
+      lastDerivedPathIdx =
+        self.userDb.getSetting(int, SettingsCol.LatestDerivedPath, 0)
+      loadedAccount = ?self.accountsGenerator.loadAccount(
+        address.get.parseAddress, password, dir)
+      newIdx = lastDerivedPathIdx + 1
+      path = fmt"{PATH_WALLET_ROOT}/{newIdx}"
+      walletAccount = ?self.storeDerivedAccount(loadedAccount.id, KeyPath path,
+        name, password, dir, AccountType.Generated)
+
+    self.userDb.saveSetting(SettingsCol.LatestDerivedPath, newIdx)
+
+    return WalletAccountResult.ok(walletAccount)
+  except SettingDbError as e:
+    return WalletAccountResult.err errorMsgSetting & e.msg
+  except StatusApiError as e:
+    return WalletAccountResult.err errorMsgSetting & e.msg
+  except UnpackError as e:
+    return WalletAccountResult.err errorMsgSetting & e.msg
+  except ValueError as e:
+    return WalletAccountResult.err errorMsgParse & e.msg
 
 proc storeImportedWalletAccount(self: StatusObject, privateKey: SkSecretKey,
-  name, password, dir: string, accountType: AccountType): WalletAccountResult =
+  name, password, dir: string, accountType: AccountType): WalletAccountResult
+  {.raises: [].} =
```

Comment:
Is it important for some procs to basically undo what `{.push raises: [Defect].}` does at the top of the module?

Reply:
It means that there is no chance for that proc to raise a `Defect` based on the compiler's analysis. Because `{.push raises: [Defect].}` effectively adds a `{.raises: [Defect].}` to each proc/func/method/template/macro in the module, if there is a proc/func/method/template/macro that does not raise `Defect`, the `raises` pragma needs to be explicitly overridden for the compiler to understand.

Reply:
> It means that there is no chance for that proc to raise a Defect based on the compiler's analysis. 

This unfortunately is not quite so - in nim 1.4, `raises: []` means that `Defect` can still be raised, while in nim 1.2 this transition was done only partially and some `Defect` trigger a compile-time error and some don't - we use `push raises: [Defect]` to prepare the code for nim 1.4 where it'll be replaced with `push raises: []` throughout, because this is the semantic that most closely matches where nim is heading.

Reply:
Makes sense Jacek, thanks so much for the info.